packages/provider/src/multi-wallet.ts

```diff
@@ -0,0 +1,271 @@
+import { Signer as AbstractSigner, Contract, ethers } from 'ethers'
+import { TransactionResponse, JsonRpcProvider } from 'ethers/providers'
+import { Wallet } from './wallet'
+import { NetworkConfig, ArcadeumWalletConfig, ArcadeumContext, Transactionish } from './types'
+import { Arrayish, BigNumberish } from 'ethers/utils'
+import { abi as mainModuleUpgradableAbi } from './abi/mainModuleUpgradable'
+import { abi as requireUtilsAbi } from './abi/requireUtils'
+import { isConfig } from './utils'
+import { NotEnoughSigners } from './errors'
+
+// TODO: Add more details to network
+// authChain, mainNetwork, etc
+export type NetworkWallet = {
+  wallet: Wallet,
+  network: NetworkConfig
+}
+
+export type MultiWalletParams = {
+  context: ArcadeumContext,
+  initialConfig: ArcadeumWalletConfig,
+  signers: AbstractSigner[],
+  networks: NetworkConfig[]
+}
+
+type FullConfig = {
+  threshold: Threshold[],
+  signers: Signer[]
+}
+
+type Threshold = {
+  chaind: number,
+  weight: number
+}
+
+type Signer = {
+  address: string,
+  networks: {
+    chaind: number,
+    weight: number
+  }[]
+}
+
+export class MultiWallet extends AbstractSigner {
+  private readonly _wallets: NetworkWallet[]
+  provider: ethers.providers.JsonRpcProvider
+
+  constructor(params: MultiWalletParams) {
+    super()
+
+    if (params.networks.length === 0) throw new Error('MultiWallet must have networks')
+
+    // Generate wallets using the initial configuration
+    this._wallets = params.networks.map((network) => {
+      let wallet = new Wallet(params.initialConfig, params.context, ...params.signers)
+      wallet.setProvider(network.rpcUrl)
+      if (network.relayer) {
+        wallet.setRelayer(network.relayer)
+      }
+      return {
+        network: network,
+        wallet: wallet
+      }
+    })
+
+    this.provider = this._wallets[0].wallet.provider
+  }
+
+  get address(): string {
+    return this._wallets[0].wallet.address
+  }
+
+  getAddress(): Promise<string> {
+    return this._wallets[0].wallet.getAddress()
+  }
+
+  async getFullConfig(): Promise<FullConfig> {
+    const allConfigs = await Promise.all(this._wallets.map(async (w) => ({ wallet: w, config: await this.currentConfig(w.wallet) })))
+    const thresholds = allConfigs.map((c) => ({ chaind: c.wallet.network.chainId, weight: c.config.threshold }))
+    const allSigerns = allConfigs.reduce((p, config) => {
+      config.config.signers.forEach((signer) => {
+        const item = p.find((c) => c.address === signer.address)
+        const netEntry = {
+          weight: signer.weight,
+          chaind: config.wallet.network.chainId
+        }
+
+        if (!item) {
+          p.push({
+            address: signer.address,
+            networks: [netEntry]
```

Comment:
why is the `weight` in the `networks` field? Seems a bit un-intuitive. Why not have three fields `address, network, weight`?

Reply:
The address may or may not exist on different networks, and on each network the address could have a different weight. Having the 3 fields works to, but at that point it's practically an array of configurations. My idea here was to try to do a "flatten" version of the configuration since that's needed to show the devices list on the wallet webapp